### PR TITLE
Fix two saving/loading bugs

### LINKED ISSUE #77: Save as... throws exception and doesn't show the dialog
Save as... on the file menu throws exceptions and doesn't show the dialog.

### LINKED ISSUE #79: Load layout gets an error and no dialog
Load layout gets an error and no dialog box.

### LINKED ISSUE #216: Loading simple save is broken
Test:
```java
public class BasicLoadTest extends ApplicationTest {

    @Override
    public void start(Stage stage) throws Exception {
        // Just here so we can run on the FX thread
        new Shuffleboard().start(stage);
    }

    @Test
    public void loadBasicFile() {
        Reader reader = new InputStreamReader(
                getClass().getResourceAsStream("/simpleshuffleboard.json"),
                StandardCharsets.UTF_8);
        JsonBuilder.forSaveFile().fromJson(reader, DashboardTabPane.class);
    }
}
```

`simpleshuffleboard.json`
```json
[
  {
    "title": "SmartDashboard",
    "widgetPane": {
      "1,1": {
        "size": [
          1,
          1
        ],
        "content": {
          "_type": "Boolean Box",
          "_source": "example://",
          "colorWhenTrue": "#7CFC00FF",
          "colorWhenFalse": "#8B0000FF"
        }
      },
      "4,3": {
        "size": [
          2,
          1
        ],
        "content": {
          "_type": "Encoder",
          "_source": "example://"
        }
      }
    }
  },
  {
    "title": "LiveWindow",
    "widgetPane": {}
  }
]

```

Exception:
```
WARNING: Couldn't load
java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:135)
	at edu.wpi.first.shuffleboard.api.sources.SourceTypes.lambda$static$1(SourceTypes.java:34)
	at edu.wpi.first.shuffleboard.api.sources.SourceType.lambda$forUri$0(SourceType.java:75)
	at edu.wpi.first.shuffleboard.api.sources.Sources.lambda$computeIfAbsent$2(Sources.java:65)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1126)
	at edu.wpi.first.shuffleboard.api.sources.Sources.computeIfAbsent(Sources.java:65)
	at edu.wpi.first.shuffleboard.api.sources.SourceType.forUri(SourceType.java:75)
	at edu.wpi.first.shuffleboard.api.sources.SourceTypes.forUri(SourceTypes.java:116)
	at edu.wpi.first.shuffleboard.api.sources.Sources.lambda$forUri$1(Sources.java:55)
	at edu.wpi.first.shuffleboard.api.sources.Sources.lambda$computeIfAbsent$2(Sources.java:65)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1126)
	at edu.wpi.first.shuffleboard.api.sources.Sources.computeIfAbsent(Sources.java:65)
	at edu.wpi.first.shuffleboard.api.sources.Sources.forUri(Sources.java:55)
	at edu.wpi.first.shuffleboard.app.json.WidgetSaver.deserialize(WidgetSaver.java:34)
	at edu.wpi.first.shuffleboard.app.json.WidgetSaver.deserialize(WidgetSaver.java:13)
	at edu.wpi.first.shuffleboard.app.json.ElementTypeAdapter.deserialize(ElementTypeAdapter.java:25)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:953)
	at com.google.gson.internal.bind.TreeTypeAdapter$GsonContextImpl.deserialize(TreeTypeAdapter.java:162)
	at edu.wpi.first.shuffleboard.app.json.WidgetPaneSaver.deserialize(WidgetPaneSaver.java:59)
	at edu.wpi.first.shuffleboard.app.json.WidgetPaneSaver.deserialize(WidgetPaneSaver.java:21)
	at edu.wpi.first.shuffleboard.app.json.ElementTypeAdapter.deserialize(ElementTypeAdapter.java:25)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:953)
	at com.google.gson.internal.bind.TreeTypeAdapter$GsonContextImpl.deserialize(TreeTypeAdapter.java:162)
	at edu.wpi.first.shuffleboard.app.json.DashboardTabPaneSaver.deserialize(DashboardTabPaneSaver.java:45)
	at edu.wpi.first.shuffleboard.app.json.DashboardTabPaneSaver.deserialize(DashboardTabPaneSaver.java:16)
	at edu.wpi.first.shuffleboard.app.json.ElementTypeAdapter.deserialize(ElementTypeAdapter.java:25)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:826)
	at edu.wpi.first.shuffleboard.app.MainWindowController.load(MainWindowController.java:360)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275)
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1771)
	at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1657)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1405)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$343(ContextMenuContent.java:1358)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:380)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:294)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$354(GlassViewEventHandler.java:416)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:415)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
```

### PR DESCRIPTION
**Pull request by nightpool**
We don't check for the existence of the SmartDashboard folder before passing it to FilePicker, which leads to an exception.

Additionally, loading empty widgets prevented the loading of the entire file. This now correctly surrounds that in a try/catch

fixes #77, fixes #79 
closes #216, closes #217

### CODE REVIEW

File: src/main/java/edu/wpi/first/shuffleboard/MainWindowController.java [hunk 1]
```diff
@@ -51,6 +37,17 @@
...
+import org.fxmisc.easybind.EasyBind;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
```
Comment (nightpool) [inline src/main/java/edu/wpi/first/shuffleboard/MainWindowController.java]: intellij, please....

**Comment by nightpool** [comment]
Tests are intermittent failures in singlekeysource

File: src/main/java/edu/wpi/first/shuffleboard/sources/SourceType.java [hunk 2]
```diff
@@ -9,7 +9,7 @@
 public enum SourceType {
 
   NONE(false, "", null),
-  STATIC(false, "example://", s -> DummySource.forTypes(DataType.forName(s)).get()),
+  STATIC(false, "example://", s -> DummySource.forTypes(DataType.forName(s)).orElse(null)),
```
Comment (SamCarlberg) [inline src/main/java/edu/wpi/first/shuffleboard/sources/SourceType.java]: Why was this changed?

Comment (nightpool) [inline src/main/java/edu/wpi/first/shuffleboard/sources/SourceType.java]: because this makes it throw the error in forURI rather then here. Looking at it now though, it could probably return a DataSource.none()

Comment (SamCarlberg) [inline src/main/java/edu/wpi/first/shuffleboard/sources/SourceType.java]: It should, `DataSource.none()` is explicitly intended to be used for "null" source values

**Comment by nightpool** [comment]
I'll resolve the conflicts and fix MainWindowController after I wrap up the first layouts.

**Comment by JLLeitschuh** [comment]
Merge conflicts need to be resolved.

**Comment by nightpool** [comment]
@SamCarlberg merged and changes addressed. Re-review?

File: api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java [hunk 3]
```diff
@@ -29,21 +29,26 @@ public SourceType getType() {
...
     }
   }
 
-  public static Optional<DummySource> forTypes(DataType... types) {
+  public static DataSource<?> forTypes(DataType... types) {
     return forTypes(ImmutableSet.copyOf(types));
   }
 
+  public static DataSource forName(String name) {
+    return DummySource.forTypes(DataTypes.getDefault().forName(name)
```
Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java]: This could be replaced with

```java
return DataTypes.getDefault().forName(name)
    .map(DummySource::forTypes)
    .orElseThrow(() -> new IllegalArgumentException("No DataType " + name));
```

File: app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java [hunk 4]
```diff
@@ -31,7 +36,11 @@ public Widget deserialize(JsonElement json, JsonDeserializationContext context)
             .orElseThrow(() -> new JsonParseException("No widget found for " + type)).get();
 
     String source = obj.get("_source").getAsString();
-    widget.setSource(Sources.getDefault().forUri(source));
+    try {
+      widget.setSource(Sources.getDefault().forUri(source));
+    } catch (RuntimeException e) {
```
Comment (SamCarlberg) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java]: This is very broad... what kind of exceptions are you expecting to be thrown, and under what circumstances?

File: api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java [hunk 5]
```diff
@@ -29,21 +29,26 @@ public SourceType getType() {
...
+      return new DummySource(type, type.getDefaultValue());
     }
   }
 
-  public static Optional<DummySource> forTypes(DataType... types) {
+  public static DataSource<?> forTypes(DataType... types) {
     return forTypes(ImmutableSet.copyOf(types));
   }
 
+  public static DataSource forName(String name) {
```
Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java]: I'd call this method `forTypeName`. Also needs documentation

File: api/src/main/java/edu/wpi/first/shuffleboard/api/util/Storage.java [hunk 6]
```diff
@@ -14,28 +19,49 @@
...
+  private static Path findOrCreate(String directory) throws IOException {
+    Path path = Paths.get(directory);
+    if (!Files.exists(path)) {
+      Files.createDirectories(path);
+    }
+
+    return path;
+  }
+
+  public static File getStorageDir() throws IOException {
```
Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/util/Storage.java]: These methods need documentation

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 7]
```diff
@@ -1,5 +1,9 @@
 package edu.wpi.first.shuffleboard.app;
 
+import de.codecentric.centerdevice.javafxsvg.SvgImageLoaderFactory;
+import it.sauronsoftware.junique.AlreadyLockedException;
+import it.sauronsoftware.junique.JUnique;
```
Comment (SamCarlberg) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: Unnecessary import reordering

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 8]
```diff
@@ -114,11 +108,6 @@ public void start(Stage primaryStage) throws IOException {
    * @throws IOException if the plugin directory could not be read
    */
   private void loadPluginsFromDir() throws IOException {
-    Path pluginPath = Paths.get(Storage.PLUGINS_DIR);
-    if (!Files.exists(pluginPath)) {
-      Files.createDirectories(pluginPath);
-    }
-    PluginLoader.getDefault().loadAllJarsFromDir(pluginPath);
+    PluginLoader.getDefault().loadAllJarsFromDir(Storage.getPluginPath());
```
Comment (SamCarlberg) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: This whole method could just get inlined now it's just a single line of code.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java [hunk 9]
```diff
@@ -29,21 +29,26 @@ public SourceType getType() {
    * If no example source value could be found, then None is returned instead.
    */
   @SuppressWarnings("unchecked")
-  public static Optional<DummySource> forTypes(Set<DataType> types) {
+  public static DataSource<?> forTypes(Set<DataType> types) {
```
Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java]: 👍

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java (hunk 7 above)
Comment (nightpool) [reply app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: it was done automatically by intellij

File: app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java (hunk 4 above)
Comment (nightpool) [reply app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java]: hmm it looks like this API has changed since I wrote this code. forUri used to throw a runtime exception if the protocol wasn't registered.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java (hunk 3 above)
Comment (nightpool) [reply api/src/main/java/edu/wpi/first/shuffleboard/api/sources/DummySource.java]: seems clearer this way?

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java (hunk 8 above)
Comment (nightpool) [reply app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: Didn't want to remove the large amount of documentation here. Down to do what you think is appropriate though.

Comment (SamCarlberg) [reply app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: TBH that documentation should be in `Plugin` so third-party API consumers can see it. I shouldn't have put it here in the first place.

File: api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java [hunk 10]
```diff
@@ -61,12 +61,12 @@ public void testBooleanToBytes() {
   @Test
   @Disabled("Adapters will be moved")
```
Comment (JLLeitschuh) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: Looks like you've updated this test. Should it still be disabled?

Comment (nightpool) [reply api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: it was only updated because I changed the signature of a method it depends on.

Comment (JLLeitschuh) [reply api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: Does the test still fail?

**Comment by AustinShalit** [comment]
Please rebase with the latest master.  frcjenkins **must** pass now.

File: api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java (hunk 10 above)
Comment (JLLeitschuh) [reply api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: Ping!

Comment (nightpool) [reply api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: while I appreciate the thought, this is out of scope for this PR

File: app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java [hunk 11]
```diff
@@ -5,13 +5,15 @@
...
 import com.google.gson.JsonSerializationContext;
+
 import edu.wpi.first.shuffleboard.api.sources.Sources;
 import edu.wpi.first.shuffleboard.api.widget.Widget;
 import edu.wpi.first.shuffleboard.api.widget.Components;
 import javafx.beans.property.Property;
 
 @AnnotatedTypeAdapter(forType = Widget.class)
 public class WidgetSaver implements ElementTypeAdapter<Widget> {
+
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/json/WidgetSaver.java]: need to add back in the SourceTypeException or a check for None sources.

File: api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java (hunk 10 above)
Comment (JLLeitschuh) [reply api/src/test/java/edu/wpi/first/shuffleboard/api/sources/recording/SerializationTest.java]: I mean if the test no longer fails it makes sense to re-enable it.

**Comment by bradamiller** [comment]
Is this still waiting for review and still relevant? Or is there something not yet finished?

**Comment by nightpool** [comment]
gonna open a new PR
